src/compiler/checker.ts

```diff
@@ -2225,21 +2227,20 @@ namespace ts {
         // An external module with an 'export =' declaration resolves to the target of the 'export =' declaration,
         // and an external module with no 'export =' declaration resolves to the module itself.
         function resolveExternalModuleSymbol(moduleSymbol: Symbol, dontResolveAlias?: boolean): Symbol {
-            return moduleSymbol && getMergedSymbol(resolveSymbol(getCommonJsExportEquals(moduleSymbol), dontResolveAlias)) || moduleSymbol;
+            return moduleSymbol && getMergedSymbol(getCommonJsExportEquals(resolveSymbol(moduleSymbol.exports.get(InternalSymbolName.ExportEquals), dontResolveAlias), moduleSymbol)) || moduleSymbol;
         }
 
-        function getCommonJsExportEquals(moduleSymbol: Symbol): Symbol {
-            const exported = moduleSymbol.exports.get(InternalSymbolName.ExportEquals);
-            if (!exported || !exported.exports || moduleSymbol.exports.size === 1) {
+        function getCommonJsExportEquals(exported: Symbol, moduleSymbol: Symbol): Symbol {
+            if (!exported || moduleSymbol.exports.size === 1) {
                 return exported;
             }
             const merged = cloneSymbol(exported);
-            moduleSymbol.exports.forEach((s, name) => {
-                if (name === InternalSymbolName.ExportEquals) return;
-                if (!merged.exports.has(name)) {
-                    merged.exports.set(name, s);
-                }
-            });
+            if (merged.exports === undefined) {
+                merged.flags = merged.flags | SymbolFlags.ValueModule;
+                merged.exports = createSymbolTable();
+            }
+            mergeSymbolTable(merged.exports, moduleSymbol.exports);

```

Comment:
So this is subtly different - if you somehow bind an `export=` and normal exports, this will merge the normal export with an `export=` member of the same name (creating a merge symbol in the process), whereas previously the first symbol in won.

Reply:
Which, I guess in JS might actually be possible, since you can do
```js
module.exports.x = 2;
class Foo {};
module.exports = Foo; // should hide all prior assignments
```

Reply:
It is the exported thing that has to have a property named `export=` for this to manifest, not the module. In what cases can the exported thing have `export=`? I can't think of any. Namespaces don't exist in JS, and that's the only construct I am not sure about.

(Your example is a real problem, it's just not related to this change.)

Reply:
After much discussion and experimentation, it looks like I can move this merging code (added in #23228) into the binder as long as we correctly handle conflicts between property assignments names and contents of the export assignment (eg `module.exports.x = ...` and `module.exports = { x: ... }`.

However, that change is not trivial, so I want to put it in a separate PR so it doesn't block webpack PR webpack/webpack#7031 that is waiting on this PR.